packages/material-ui-lab/src/useAutocomplete/useAutocomplete.d.ts

```diff
@@ -18,7 +18,22 @@ export function createFilterOptions<T>(
   config?: CreateFilterOptionsConfig<T>
 ): (options: T[], state: FilterOptionsState<T>) => T[];
 
-export interface UseAutocompleteCommonProps<T> {
+export type AutocompleteFreeSoloValueMapping<T> = T extends true ? string : never;
+
+export type TValue<T, TMultiple, TDisableClearable, TFreeSolo> = TMultiple extends undefined | false
```

Comment:
The order isn't important here. Can we rather pass the props type or some type map? Otherwise it gets kind of annoying to used since TypeScript doesn't support partial type parameters.

Reply:
I can't find how to make this work. I have tried with no luck:

```diff
diff --git a/packages/material-ui-lab/src/Autocomplete/Autocomplete.d.ts b/packages/material-ui-lab/src/Autocomplete/Autocomplete.d.ts
index 03a025e82..fba715c14 100644
--- a/packages/material-ui-lab/src/Autocomplete/Autocomplete.d.ts
+++ b/packages/material-ui-lab/src/Autocomplete/Autocomplete.d.ts
@@ -45,13 +45,8 @@ export interface AutocompleteRenderInputParams {
   inputProps: object;
 }

-export interface AutocompleteProps<
-  T,
-  Multiple extends boolean | undefined,
-  DisableClearable extends boolean | undefined,
-  FreeSolo extends boolean | undefined
->
-  extends UseAutocompleteProps<T, Multiple, DisableClearable, FreeSolo>,
+export interface AutocompleteProps<T>
+  extends UseAutocompleteProps<T>,
     StandardProps<
       React.HTMLAttributes<HTMLDivElement>,
       AutocompleteClassKey,
@@ -217,9 +212,4 @@ export type AutocompleteClassKey =
  *
  * - [Autocomplete API](https://material-ui.com/api/autocomplete/)
  */
-export default function Autocomplete<
-  T,
-  Multiple extends boolean | undefined = undefined,
-  DisableClearable extends boolean | undefined = undefined,
-  FreeSolo extends boolean | undefined = undefined
->(props: AutocompleteProps<T, Multiple, DisableClearable, FreeSolo>): JSX.Element;
+export default function Autocomplete<T>(props: AutocompleteProps<T>): JSX.Element;
diff --git a/packages/material-ui-lab/src/Autocomplete/Autocomplete.spec.tsx b/packages/material-ui-lab/src/Autocomplete/Autocomplete.spec.tsx
index 927ea1de3..bfcb510fe 100644
--- a/packages/material-ui-lab/src/Autocomplete/Autocomplete.spec.tsx
+++ b/packages/material-ui-lab/src/Autocomplete/Autocomplete.spec.tsx
@@ -1,20 +1,10 @@
 import { Autocomplete, AutocompleteProps } from '@material-ui/lab';

-interface MyAutocomplete<
-  T,
-  Multiple extends boolean | undefined,
-  DisableClearable extends boolean | undefined,
-  FreeSolo extends boolean | undefined
-> extends AutocompleteProps<T, Multiple, DisableClearable, FreeSolo> {
+interface MyAutocomplete<T> extends AutocompleteProps<T> {
   myProp?: string;
 }

-function MyAutocomplete<
-  T,
-  Multiple extends boolean | undefined,
-  DisableClearable extends boolean | undefined,
-  FreeSolo extends boolean | undefined
->(props: MyAutocomplete<T, Multiple, DisableClearable, FreeSolo>) {
+function MyAutocomplete<T>(props: MyAutocomplete<T>) {
   return <Autocomplete {...props} />;
 }

diff --git a/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.d.ts b/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.d.ts
index 697d83f66..a8e459c6f 100644
--- a/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.d.ts
+++ b/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.d.ts
@@ -20,17 +20,21 @@ export function createFilterOptions<T>(

 export type AutocompleteFreeSoloValueMapping<T> = T extends true ? string : never;

-export type Value<T, Multiple, DisableClearable, FreeSolo> = Multiple extends undefined | false
-  ? DisableClearable extends true
-    ? NonNullable<T | AutocompleteFreeSoloValueMapping<FreeSolo>>
-    : T | null | AutocompleteFreeSoloValueMapping<FreeSolo>
-  : Array<T | AutocompleteFreeSoloValueMapping<FreeSolo>>;
+export interface ValuePropsType {
+  disableClearable?: boolean;
+  freeSolo?: boolean;
+  multiple?: boolean;
+}
+
+export type Value<T, ValueProps extends ValuePropsType> = ValueProps['multiple'] extends undefined | false
+  ? ValueProps['disableClearable'] extends true
+    ? NonNullable<T | AutocompleteFreeSoloValueMapping<ValueProps['freeSolo']>>
+    : T | null | AutocompleteFreeSoloValueMapping<ValueProps['freeSolo']>
+  : Array<T | AutocompleteFreeSoloValueMapping<ValueProps['freeSolo']>>;

 export interface UseAutocompleteProps<
   T,
-  Multiple extends boolean | undefined,
-  DisableClearable extends boolean | undefined,
-  FreeSolo extends boolean | undefined
+  ValueProps extends ValuePropsType = {}
 > {
   /**
    * If `true`, the portion of the selected suggestion that has not been typed by the user,
@@ -81,7 +85,7 @@ export interface UseAutocompleteProps<
   /**
    * If `true`, the input can't be cleared.
    */
-  disableClearable?: DisableClearable;
+  disableClearable?: ValueProps['disableClearable'];
   /**
    * If `true`, the popup won't close when a value is selected.
    */
@@ -109,7 +113,7 @@ export interface UseAutocompleteProps<
   /**
    * If `true`, the Autocomplete is free solo, meaning that the user input is not bound to provided options.
    */
-  freeSolo?: FreeSolo;
+  freeSolo?: ValueProps['freeSolo'];
   /**
    * Used to determine the disabled state for a given option.
    *
@@ -219,18 +223,18 @@ export interface UseAutocompleteProps<
   /**
    * If `true`, `value` must be an array and the menu will support multiple selections.
    */
-  multiple?: Multiple;
+  multiple?: ValueProps['multiple'];
   /**
    * The value of the autocomplete.
    *
    * The value must have reference equality with the option in order to be selected.
    * You can customize the equality behavior with the `getOptionSelected` prop.
    */
-  value?: Value<T, Multiple, DisableClearable, FreeSolo>;
+  value?: Value<T, ValueProps>;
   /**
    * The default input value. Use when the component is not controlled.
    */
-  defaultValue?: Value<T, Multiple, DisableClearable, FreeSolo>;
+  defaultValue?: Value<T, ValueProps>;
   /**
    * Callback fired when the value changes.
    *
@@ -240,7 +244,7 @@ export interface UseAutocompleteProps<
    */
   onChange?: (
     event: React.ChangeEvent<{}>,
-    value: Value<T, Multiple, DisableClearable, FreeSolo>,
+    value: Value<T, ValueProps>,
     reason: AutocompleteChangeReason,
     details?: AutocompleteChangeDetails<T>
   ) => void;
@@ -262,11 +266,9 @@ export type AutocompleteInputChangeReason = 'input' | 'reset' | 'clear';

 export default function useAutocomplete<
   T,
-  Multiple extends boolean | undefined = undefined,
-  DisableClearable extends boolean | undefined = undefined,
-  FreeSolo extends boolean | undefined = undefined
+  ValueProps extends ValuePropsType = {}
 >(
-  props: UseAutocompleteProps<T, Multiple, DisableClearable, FreeSolo>
+  props: UseAutocompleteProps<T, ValueProps>
 ): {
   getRootProps: () => {};
   getInputProps: () => {};
@@ -278,7 +280,7 @@ export default function useAutocomplete<
   getOptionProps: ({ option, index }: { option: T; index: number }) => {};
   id: string;
   inputValue: string;
-  value: Value<T, Multiple, DisableClearable, FreeSolo>;
+  value: Value<T, ValueProps>;
   dirty: boolean;
   popupOpen: boolean;
   focused: boolean;
```

Reply:
@justinabrahms By any chance, will you have the time to look at this problem? I think that it's the last blocker before we can merge the pull request. There is a lot of goodness in this pull request, I'm excited about the idea we can release it :).

Reply:
@oliviertassinari sure, I'll try to take a look today. Had a very busy last week.

Reply:
I don't find this important anymore since I don't think you pass these type parameters explicitly anyway. If this comes up more often we can always switch (because alpha). @kanoshin feel free to mark this as resolved unless you think this is important.

Reply:
@eps1lon spent some time thinking and I'm not sure that I understand the suggestion (and it's feasibility). Since all `TMultiple`, `TDisableClearable`, `TFreeSolo` are generic types it's still required to declare them somewhere to use in type mapping. Could you elaborate more?

Something like this won't work because TS would use undefined type without explicitly passed generics:
```
interface ValueProps<Multiple = undefined, DisableClearable = undefined, FreeSolo = undefined> {
  disableClearable?: DisableClearable;
  freeSolo?: FreeSolo;
  multiple?: Multiple;
}

export type Value<T, TValueProps extends ValueProps> = TValueProps['multiple'] extends undefined | false
  ? TValueProps['disableClearable'] extends true
    ? NonNullable<T | AutocompleteFreeSoloValueMapping<TValueProps['freeSolo']>>
    : T | null | AutocompleteFreeSoloValueMapping<TValueProps['freeSolo']>
  : Array<T | AutocompleteFreeSoloValueMapping<TValueProps['freeSolo']>>;
```

>I don't think you pass these type parameters explicitly anyway.

Currently interfaces don't have default values because I wanted to avoid duplicated code, but there are default values on component and `useAutocomplete` hook level to avoid having to declare them explicitly in the userland as you noticed.

Reply:
Yeah let's revisit this once we actually need it.